Model/IcingVersion.php

```diff
@@ -171,18 +251,30 @@ public function saveVersion($data, $settings = array()) {
 			'model_id' => $data['model_id'],
 		);
 
+		// Only force extra condition if we have soft_deleting on.
+		if (isset($settings['soft_delete']) && $settings['soft_delete']) {
+			$conditions['is_soft_delete'] = 0;
+		}
+
 		// check if we should prune off old records for this model+id
 		//   (limited by settings - versions count)
 		if (!empty($settings['versions']) && is_numeric($settings['versions']) && $settings['versions'] > 0) {
 			$count = $this->find('count', compact('conditions'));
-			while ($count >= $settings['versions']) {
-				$oldest = $this->find('first', array(
+
+			if ($count >= $settings['versions']) {
+				$limit = $count - $settings['versions'] + 1; // removing current count minus setting (plus one to make room for this new version)
+				$versions_to_remove = $this->find('list', array(
 					'fields' => array('id'),
 					'conditions' => $conditions,
-					'order' => array("{$this->alias}.created ASC")
+					'order' => array("{$this->alias}.created ASC"),
+					'limit' => $limit,
 				));
-				$this->delete($oldest[$this->alias]['id']);
-				$count = $this->find('count', compact('conditions'));
+
+				if (isset($settings['soft_delete']) && $settings['soft_delete']) {
```

Comment:
Should we extract this into its own method with early returning to drop the else?

Reply:
Can’t drop the else as there is no default behavior. I’m open to suggestions on what you’d think removing the else might look like. Perhaps I’m just not seeing it. 

I do like the idea of extracting the condition of soft delete setting. It’s used often.

Reply:
I was thinking something along the lines of this:
`$this->handleDeletions($settings['soft_delete'], $versions_to_remove);`

Then that function would do something like (note, could flip to whatever you think the default behavior should be)
`public function handleDeletions(bool $softDelete, string $versionsToRemove): bool {`
`    if($softDelete) {`
`        return $this->softDeleteAll($versions_to_remove);`
`    }`
`    return $this->hardDeleteAll($versions_to_remove);`
`}`